File: TvNavigationLayout-java/app/src/main/java/com/hxxdashai/android/navigation/widget/NavigationLinearLayout.java
package com.hxxdashai.android.navigation.widget;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.Rect;
import android.support.v4.view.ViewCompat;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.SparseIntArray;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.hxxdashai.android.navigation.R;
import com.hxxdashai.android.navigation.util.SoundUtil;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Mr.T on 2018/4/25.
 */

public class NavigationLinearLayout extends LinearLayout {

    public static final int STATE_NO_SELECT = 666;//默认状态
    public static final int STATE_HAS_SELECT_NO_fOCUS = 667;//选中无焦点
    public static final int STATE_HAS_SELECT_HAS_fOCUS = 668;//选中有焦点

    public static final String MODE_SAME = "same";//item固定宽模式
    public static final String MODE_SELF = "self";//item自适应宽模式

    private float fontSize;//字体大小
    private float enlargeRate;//放大倍率
    private int fontColorNormal;//默认字体颜色
    private int fontColorSelect;//选中字体颜色
    private int fontColorLight;//选中字体发光颜色
    private int defaultPos;//默认选中的pos
    private String orderMode;//item排列模式，"same":固定宽模式，"self":自适应宽模式
    private int itemSpace;//"same":item宽度，"self":item距左或右宽度(实际每个item间距是两个itemSpace值)

    private List<String> mDataList = new ArrayList();
    private SparseIntArray mToLeftMap = new SparseIntArray();
    private int mNowPos = -1;
    private NavigationListener mNavigationListener;
    private NavigationCursorView mNavigationCursorView;

    public NavigationLinearLayout(Context context) {
        this(context, null);
    }

    public NavigationLinearLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public NavigationLinearLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        if (attrs != null) {
            TypedArray attributes = context.obtainStyledAttributes(attrs, R.styleable.home_Navigation);
            fontSize = attributes.getDimension(R.styleable.home_Navigation_home_FontSize, 30.0F);
            enlargeRate = attributes.getFloat(R.styleable.home_Navigation_home_EnlargeRate, 1.1F);
            fontColorNormal = attributes.getColor(R.styleable.home_Navigation_home_FontColorNormal, Color.WHITE);
            fontColorSelect = attributes.getColor(R.styleable.home_Navigation_home_FontColorSelect, Color.BLUE);
            fontColorLight = attributes.getColor(R.styleable.home_Navigation_home_FontColorLight, Color.RED);
            defaultPos = attributes.getInteger(R.styleable.home_Navigation_home_DefaultPos, 0);
            String mode = attributes.getString(R.styleable.home_Navigation_home_OrderMode);
            if (TextUtils.isEmpty(mode)) mode = "self";
            orderMode = mode;
            itemSpace = attributes.getDimensionPixelSize(R.styleable.home_Navigation_home_ItemSpace, 10);
            attributes.recycle();
        }
        setFocusable(true);
    }

    /**
     * 设置数据，数据改变后需重新调用(导航栏编辑等功能)
     */
    public void setDataList(List data) {
        mDataList = data;
        initView();
    }

    /**
     * 监听导航事件
     */
    public void setNavigationListener(NavigationListener listener) {
        mNavigationListener = listener;
        if (mNavigationListener != null) {
            mNavigationListener.onNavigationChange(mNowPos, KeyEvent.KEYCODE_DPAD_LEFT);
        }
    }

    /**
     * 导航光标
     */
    public void setNavigationCursorView(NavigationCursorView view) {
        mNavigationCursorView = view;
        int left = mToLeftMap.get(mNowPos);
        if (left != 0 && mNavigationCursorView != null) {
            mNavigationCursorView.fsatJumpTo(left);
        }
    }

    private void initView() {
        if (mToLeftMap.size() != 0) {
            mToLeftMap.clear();
        }
        if (mDataList.size() > getChildCount()) {
            do {
                addView(getItemView());
            } while (mDataList.size() > getChildCount());
        } else if (mDataList.size() < getChildCount()) {
            do {
                removeViewAt(getChildCount() - 1);
            } while (mDataList.size() < getChildCount());
        }
        if (mNowPos != -1 && mNowPos < getChildCount()) {
            changeItemState(mNowPos, STATE_NO_SELECT);
        }
        for (int i = 0; i < getChildCount() - 1; i++) {
            final int finalI = i;
            final TextView child = (TextView) getChildAt(i);
            child.setText(mDataList.get(i));
            child.getViewTreeObserver().addOnGlobalLayoutListener((new ViewTreeObserver.OnGlobalLayoutListener() {
                public void onGlobalLayout() {
                    child.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                    int left = child.getWidth() / 2 + child.getLeft() + getLeft();//每个item中点到父布局左边的距离
                    mToLeftMap.put(finalI, left);
                    if (defaultPos == finalI) {
                        mNowPos = defaultPos;
                        changeItemState(mNowPos, isFocused() ? STATE_HAS_SELECT_HAS_fOCUS : STATE_HAS_SELECT_NO_fOCUS);
                        if (mNavigationCursorView != null) {
                            mNavigationCursorView.fsatJumpTo(left);
                        }
                        if (mNavigationListener != null) {
                            mNavigationListener.onNavigationChange(mNowPos, KeyEvent.KEYCODE_DPAD_LEFT);
                        }
                    }
                }
            }));
        }
    }

    private void changeItemState(int pos, int state) {
        View child = getChildAt(pos);
        if (child != null) {
            switch (state) {
                case STATE_NO_SELECT:
                    //if (child.scaleX != 1f)//TODO BUG
                    ViewCompat.animate(child).scaleX(1.0F).scaleY(1.0F).translationZ(0.0F).start();
                    ((TextView) child).setShadowLayer(0.0F, 0.0F, 0.0F, fontColorLight);
                    child.setSelected(false);
                    break;
                case STATE_HAS_SELECT_NO_fOCUS:
                    if (child.getScaleX() != 1.0F)
                        ViewCompat.animate(child).scaleX(1.0F).scaleY(1.0F).translationZ(0.0F).start();
                    if (!child.isSelected()) {
                        ((TextView) child).setShadowLayer(25.0F, 0.0F, 0.0F, fontColorLight);
                        child.setSelected(true);
                    }
                    break;
                case STATE_HAS_SELECT_HAS_fOCUS:
                    ViewCompat.animate(child).scaleX(enlargeRate).scaleY(enlargeRate).translationZ(0.0F).start();
                    if (!child.isSelected()) {
                        ((TextView) child).setShadowLayer(25.0F, 0.0F, 0.0F, fontColorLight);
                        child.setSelected(true);
                    }
                    break;
            }
        }
    }

    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_DOWN) {
            switch (keyCode) {
                case KeyEvent.KEYCODE_DPAD_LEFT:
                    if (mNowPos > 0) {
                        changeItemState(mNowPos, STATE_NO_SELECT);
                        mNowPos--;
                        changeItemState(mNowPos, STATE_HAS_SELECT_HAS_fOCUS);
                        int left = mToLeftMap.get(mNowPos);
                        if (left != 0 && mNavigationCursorView != null)
                            mNavigationCursorView.jumpTo(left);
                        if (mNavigationListener != null) {
                            mNavigationListener.onNavigationChange(mNowPos, keyCode);
                        }
                    }
                    SoundUtil.playClickSound(this);
                    return true;
                case KeyEvent.KEYCODE_DPAD_RIGHT:
                    if (mNowPos < getChildCount() - 1) {
                        changeItemState(mNowPos, STATE_NO_SELECT);
                        mNowPos++;
                        changeItemState(mNowPos, STATE_HAS_SELECT_HAS_fOCUS);
                        int left = mToLeftMap.get(mNowPos);
                        if (left != 0 && mNavigationCursorView != null)
                            mNavigationCursorView.jumpTo(left);
                        if (mNavigationListener != null) {
                            mNavigationListener.onNavigationChange(mNowPos, keyCode);
                        }
                    }
                    SoundUtil.playClickSound(this);
                    return true;
                case KeyEvent.KEYCODE_DPAD_UP:
                case KeyEvent.KEYCODE_DPAD_DOWN:
                    if (mNavigationListener != null) {
                        mNavigationListener.onNavigationChange(mNowPos, keyCode);
                    }
                    break;
                case KeyEvent.KEYCODE_MENU:
                    if (mNavigationListener != null) {
                        mNavigationListener.onNavigationChange(mNowPos, keyCode);
                    }
                    return true;
            }
        }
        return super.onKeyDown(keyCode, event);
    }

    protected void onFocusChanged(boolean gainFocus, int direction, Rect previouslyFocusedRect) {
        changeItemState(mNowPos, gainFocus ? STATE_HAS_SELECT_HAS_fOCUS : STATE_HAS_SELECT_NO_fOCUS);
        if (mNavigationCursorView != null)
            mNavigationCursorView.setVisibility(gainFocus ? View.VISIBLE : View.INVISIBLE);
        super.onFocusChanged(gainFocus, direction, previouslyFocusedRect);
    }

    private TextView getItemView() {
        int[][] states = {{android.R.attr.state_selected}, new int[0]};
        int[] colors = new int[]{fontColorSelect, fontColorNormal};
        ColorStateList colorStateList = new ColorStateList(states, colors);
        TextView textView = new TextView(getContext());
        textView.setTextSize(fontSize);
        textView.setTextColor(colorStateList);
        textView.setIncludeFontPadding(false);
        switch (orderMode) {
            case MODE_SAME:
                LayoutParams layoutParams = new LayoutParams(itemSpace, LayoutParams.WRAP_CONTENT);
                textView.setLayoutParams(layoutParams);
                textView.setGravity(Gravity.CENTER);
                break;
            case MODE_SELF:
                textView.setPadding(itemSpace, 0, itemSpace, 0);
        }
        return textView;
    }

    public void jumpTo(int keyCode) {
        if (mNowPos > 0 && keyCode == KeyEvent.KEYCODE_DPAD_LEFT || mNowPos < getChildCount() - 1 && keyCode == KeyEvent.KEYCODE_DPAD_RIGHT) {
            changeItemState(mNowPos, STATE_NO_SELECT);
            if (keyCode == KeyEvent.KEYCODE_DPAD_LEFT)
                mNowPos--;
            else
                mNowPos++;
            changeItemState(mNowPos, STATE_HAS_SELECT_NO_fOCUS);
            int left = mToLeftMap.get(mNowPos);
            if (left != 0 && mNavigationCursorView != null)
                mNavigationCursorView.jumpTo(left);
            if (mNavigationListener != null) {
                mNavigationListener.onNavigationChange(mNowPos, keyCode);
            }
        }
    }

    public interface NavigationListener {

        void onNavigationChange(int pos, int keyCode);
    }
}

File: app/src/main/java/com/hxxdashai/android/navigation/widget/NavigationLinearLayout.kt
package com.hxxdashai.android.navigation.widget

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.Rect
import android.support.v4.view.ViewCompat
import android.util.AttributeSet
import android.view.Gravity
import android.view.KeyEvent
import android.view.View
import android.view.ViewTreeObserver
import android.widget.LinearLayout
import android.widget.TextView
import com.hxxdashai.android.navigation.R
import com.hxxdashai.android.navigation.util.SoundUtil

/**
 * Created by Mr.T on 2018/3/27.
 */
class NavigationLinearLayout @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : LinearLayout(context, attrs, defStyleAttr) {

    companion object {
        const val STATE_NO_SELECT = 666//默认状态
        const val STATE_HAS_SELECT_NO_fOCUS = 667//选中无焦点
        const val STATE_HAS_SELECT_HAS_fOCUS = 668//选中有焦点

        const val MODE_SAME = "same"//item固定宽模式
        const val MODE_SELF = "self"//item自适应宽模式
    }

    private var fontSize: Float = 0.0F//字体大小
    private var enlargeRate: Float = 0.0f//放大倍率
    private var fontColorNormal: Int = 0//默认字体颜色
    private var fontColorSelect: Int = 0//选中字体颜色
    private var fontColorLight: Int = 0//选中字体发光颜色
    private var defaultPos: Int = 0//默认选中的pos
    private var orderMode: String = ""//item排列模式，"same":固定宽模式，"self":自适应宽模式
    private var itemSpace: Int = 0//"same":item宽度，"self":item距左或右宽度(实际每个item间距是两个itemSpace值)

    init {
        if (attrs != null) {
            val attributes = context.obtainStyledAttributes(attrs, R.styleable.home_Navigation)
            fontSize = attributes.getDimension(R.styleable.home_Navigation_home_FontSize, 30f)
            enlargeRate = attributes.getFloat(R.styleable.home_Navigation_home_EnlargeRate, 1.1f)
            fontColorNormal = attributes.getColor(R.styleable.home_Navigation_home_FontColorNormal, Color.WHITE)
            fontColorSelect = attributes.getColor(R.styleable.home_Navigation_home_FontColorSelect, Color.BLUE)
            fontColorLight = attributes.getColor(R.styleable.home_Navigation_home_FontColorLight, Color.RED)
            defaultPos = attributes.getInteger(R.styleable.home_Navigation_home_DefaultPos, 0)
            orderMode = attributes.getString(R.styleable.home_Navigation_home_OrderMode) ?: MODE_SELF
            itemSpace = attributes.getDimensionPixelSize(R.styleable.home_Navigation_home_ItemSpace, 10)
            attributes.recycle()
        }
        isFocusable = true
    }

    /**
     * 设置数据，数据改变后需重新调用(导航栏编辑等功能)
     */
    var mDataList: MutableList<String> = ArrayList()
        set(value) {
            field = value
            initView()
        }
    private var mToLeftMap: MutableMap<Int, Int> = HashMap()//每个item中点到父布局左边的距离
    var mNowPos: Int = -1
    /**
     * 监听导航事件
     */
    var mNavigationListener: NavigationListener? = null
        set(value) {
            field = value
            field?.onNavigationChange(mNowPos, KeyEvent.KEYCODE_DPAD_LEFT)
        }
    /**
     * 导航光标
     */
    var mNavigationCursorView: NavigationCursorView? = null
        set(value) {
            field = value
            mToLeftMap[mNowPos]?.let { field?.fsatJumpTo(it) }
        }

    private fun initView() {
        if (mToLeftMap.isNotEmpty()) mToLeftMap.clear()//还原状态
        if (mDataList.size > childCount) {
            do {
                addView(getItemView())
            } while (mDataList.size > childCount)
        } else if (mDataList.size < childCount) {
            do {
                removeViewAt(childCount - 1)
            } while (mDataList.size < childCount)
        }
        if (mNowPos != -1 && mNowPos < childCount) changeItemState(mNowPos, STATE_NO_SELECT)//还原状态
        for (i in 0..(childCount - 1)) {
            val child = getChildAt(i) as TextView
            child.text = mDataList[i]
            child.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
                override fun onGlobalLayout() {
                    child.viewTreeObserver.removeOnGlobalLayoutListener(this)
                    mToLeftMap[i] = child.width / 2 + child.left + this@NavigationLinearLayout.left//每个item中点到父布局左边的距离
                    if (defaultPos == i) {//TODO 如果编辑导航后不要重置pos，可根据实际修改逻辑
                        mNowPos = defaultPos//默认要展示的pos
                        changeItemState(mNowPos, if (this@NavigationLinearLayout.isFocused) STATE_HAS_SELECT_HAS_fOCUS else STATE_HAS_SELECT_NO_fOCUS)//修改默认要展示的pos的状态
                        mToLeftMap[mNowPos]?.let { mNavigationCursorView?.fsatJumpTo(it) }//移动光标
                        mNavigationListener?.onNavigationChange(mNowPos, KeyEvent.KEYCODE_DPAD_LEFT)//展示内容数据，仅仅展示数据，写左右都没问题
                    }
                }
            })
        }
    }

    private fun changeItemState(pos: Int, state: Int) {
        val child = getChildAt(pos)
        if (child != null)
            when (state) {
                STATE_NO_SELECT -> {
                    //if (child.scaleX != 1f) ViewCompat.animate(child).scaleX(1f).scaleY(1f).translationZ(0f).start()//TODO BUG
                    ViewCompat.animate(child).scaleX(1f).scaleY(1f).translationZ(0f).start()
                    (child as TextView).setShadowLayer(0f, 0f, 0f, fontColorLight)
                    child.isSelected = false
                }
                STATE_HAS_SELECT_NO_fOCUS -> {
                    if (child.scaleX != 1f) ViewCompat.animate(child).scaleX(1f).scaleY(1f).translationZ(0f).start()
                    if (!child.isSelected) {
                        (child as TextView).setShadowLayer(25f, 0f, 0f, fontColorLight)
                        child.isSelected = true
                    }
                }
                STATE_HAS_SELECT_HAS_fOCUS -> {
                    ViewCompat.animate(child).scaleX(enlargeRate).scaleY(enlargeRate).translationZ(0f).start()
                    if (!child.isSelected) {
                        (child as TextView).setShadowLayer(25f, 0f, 0f, fontColorLight)
                        child.isSelected = true
                    }
                }
            }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (event?.action == KeyEvent.ACTION_DOWN) {
            when (keyCode) {
                KeyEvent.KEYCODE_DPAD_LEFT -> {
                    if (mNowPos > 0) {
                        changeItemState(mNowPos, STATE_NO_SELECT)
                        changeItemState(--mNowPos, STATE_HAS_SELECT_HAS_fOCUS)
                        mToLeftMap[mNowPos]?.let { mNavigationCursorView?.jumpTo(it) }
                        mNavigationListener?.onNavigationChange(mNowPos, keyCode)
                    }//如果有跳出导航栏的左右事件需求可在次此处else回调出去
                    SoundUtil.playClickSound(this@NavigationLinearLayout)
                    return true//TODO 系统声音会被屏蔽掉
                }
                KeyEvent.KEYCODE_DPAD_RIGHT -> {
                    if (mNowPos < childCount - 1) {
                        changeItemState(mNowPos, STATE_NO_SELECT)
                        changeItemState(++mNowPos, STATE_HAS_SELECT_HAS_fOCUS)
                        mToLeftMap[mNowPos]?.let { mNavigationCursorView?.jumpTo(it) }
                        mNavigationListener?.onNavigationChange(mNowPos, keyCode)
                    }
                    SoundUtil.playClickSound(this@NavigationLinearLayout)
                    return true
                }
                KeyEvent.KEYCODE_DPAD_UP, KeyEvent.KEYCODE_DPAD_DOWN -> {//TODO 方向类型的事件，不想系统自动找焦点，可试试return true
                    mNavigationListener?.onNavigationChange(mNowPos, keyCode)
                }
                KeyEvent.KEYCODE_MENU -> {//TODO 非方向类型事件
                    mNavigationListener?.onNavigationChange(mNowPos, keyCode)
                    return true//TODO bug
                }
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onFocusChanged(gainFocus: Boolean, direction: Int, previouslyFocusedRect: Rect?) {
        changeItemState(mNowPos, if (gainFocus) STATE_HAS_SELECT_HAS_fOCUS else STATE_HAS_SELECT_NO_fOCUS)
        mNavigationCursorView?.visibility = if (gainFocus) View.VISIBLE else View.INVISIBLE
        super.onFocusChanged(gainFocus, direction, previouslyFocusedRect)
    }

    private fun getItemView(): TextView {
        val states = arrayOf(intArrayOf(android.R.attr.state_selected), intArrayOf())
        val colors = intArrayOf(fontColorSelect, fontColorNormal)
        val colorStateList = ColorStateList(states, colors)
        val textView = TextView(context)
        textView.textSize = fontSize
        textView.setTextColor(colorStateList)
        textView.includeFontPadding = false
        when (orderMode) {
            MODE_SAME -> {
                val layoutParams = LayoutParams(itemSpace, LayoutParams.WRAP_CONTENT)
                textView.layoutParams = layoutParams
                textView.gravity = Gravity.CENTER
            }
            MODE_SELF -> {
                textView.setPadding(itemSpace, 0, itemSpace, 0)
            }
        }
        return textView
    }

    /**
     * 导航栏目被动切换时调用
     */
    fun jumpTo(keyCode: Int) {
        if ((mNowPos > 0 && keyCode == KeyEvent.KEYCODE_DPAD_LEFT) || (mNowPos < childCount - 1 && keyCode == KeyEvent.KEYCODE_DPAD_RIGHT)) {
            changeItemState(mNowPos, STATE_NO_SELECT)
            changeItemState(if (keyCode == KeyEvent.KEYCODE_DPAD_LEFT) --mNowPos else ++mNowPos, STATE_HAS_SELECT_NO_fOCUS)
            mToLeftMap[mNowPos]?.let { mNavigationCursorView?.jumpTo(it) }
            mNavigationListener?.onNavigationChange(mNowPos, KeyEvent.KEYCODE_DPAD_LEFT)//仅仅展示数据，写左右都没问题
        }
    }

    interface NavigationListener {

        /**
         * @param pos     选中的序号
         * @param keyCode 点击的按键
         */
        fun onNavigationChange(pos: Int, keyCode: Int)
    }
}